pkg/sink/postgres.go

```diff
@@ -561,29 +543,80 @@ func (p *PGXSink) handleUpdate(ctx context.Context, m *pb.Change) (err error) {
 		}
 	}
 
-	return p.raw.ExecParams(ctx, sql.UpdateQuery(m.Schema, m.Table, sets, keys), vals, oids, fmts, fmts).Read().Err
+	p.pipeline.SendQueryParams(sql.UpdateQuery(m.Schema, m.Table, sets, keys), vals, oids, fmts, fmts)
+	return
 }
 
 const (
 	UpdateSourceSQL = "update pgcapture.sources set commit=$1,seq=$2,mid=$3,commit_ts=$4,apply_ts=now() where id=$5"
 )
 
-func (p *PGXSink) handleCommit(cp cursor.Checkpoint, commit *pb.Commit) (err error) {
-	ctx := context.Background()
-	if err = p.flushInsert(ctx); err != nil {
+func (p *PGXSink) handleCommit(cp cursor.Checkpoint, commit *pb.Commit, change source.Change) (err error) {
+	if err = p.flushInsert(); err != nil {
 		return err
 	}
-	if _, err = p.conn.Prepare(ctx, UpdateSourceSQL, UpdateSourceSQL); err != nil {
+
+	var (
+		cmt   []byte
+		seq   []byte
+		mid   []byte
+		cmtTs []byte
+		id    []byte
+	)
+
+	// pgtype does not support uint64, so we have to encode it as text
+	cmt, err = p.conn.TypeMap().Encode(0, pgtype.TextFormatCode, pgLSN(cp.LSN), nil)
+	if err != nil {
+		return err
+	}
+	seq, err = p.conn.TypeMap().Encode(pgtype.Int4OID, pgtype.BinaryFormatCode, pgInt4(int32(cp.Seq)), nil)
+	if err != nil {
 		return err
 	}
-	commitTs := pgTz(commit.CommitTime)
-	if _, err = p.conn.Exec(ctx, UpdateSourceSQL, pgLSN(cp.LSN), pgInt4(int32(cp.Seq)), cp.Data, commitTs, p.pgSrcID); err != nil {
+	mid, err = p.conn.TypeMap().Encode(pgtype.ByteaOID, pgtype.BinaryFormatCode, cp.Data, nil)
+	if err != nil {
 		return err
 	}
-	if _, err = p.conn.Exec(ctx, "commit"); err != nil {
+	cmtTs, err = p.conn.TypeMap().Encode(pgtype.TimestamptzOID, pgtype.BinaryFormatCode, pgTz(commit.CommitTime), nil)
+	if err != nil {
 		return err
 	}
-	atomic.StoreInt64(&p.replLag, time.Since(commitTs.Time).Milliseconds())
+	id, err = p.conn.TypeMap().Encode(pgtype.TextOID, pgtype.BinaryFormatCode, p.pgSrcID, nil)
+	if err != nil {
+		return err
+	}
+
+	p.pipeline.SendQueryParams(UpdateSourceSQL, [][]byte{cmt, seq, mid, cmtTs, id}, []uint32{0, pgtype.Int4OID, pgtype.ByteaOID, pgtype.TimestamptzOID, pgtype.TextOID}, []int16{pgtype.TextFormatCode, pgtype.BinaryFormatCode, pgtype.BinaryFormatCode, pgtype.BinaryFormatCode, pgtype.BinaryFormatCode}, []int16{pgtype.TextFormatCode, pgtype.BinaryFormatCode, pgtype.BinaryFormatCode, pgtype.BinaryFormatCode, pgtype.BinaryFormatCode})
+	p.currentCommit = commit
+	p.pendingCommitted = append(p.pendingCommitted, change.Checkpoint)
+	if !change.HasNext || len(p.pendingCommitted) == p.BatchTXSize {
```

Comment:
what does it exactly means if change.HasNext == ture?

Reply:
oh I should rewrite to this:
```go
if len(p.pendingCommitted) == p.BatchTXSize || !change.HasNext {
	err = p.endPipeline()
}
```
Otherwise, `pendingCommitted` will continue to accumulate and cannot be flushed.

Reply:
When `HasNext == true`, it means there is a next begin message.